rstest_macros/src/render/mod.rs

```diff
@@ -200,19 +203,32 @@ pub(crate) fn matrix(mut test: ItemFn, mut info: RsTestInfo) -> TokenStream {
                     &info,
                     &c.info,
                 )
-                .wrap_by_mod(&c.ident)
+                .map(|recur| recur.wrap_by_mod(&c.ident))
             })
-            .collect()
+            .collect::<Result<_, _>>()?
     };
 
-    test_group(test, rendered_cases)
+    Ok(test_group(test, rendered_cases))
 }
 
-fn resolve_default_test_attr(is_async: bool) -> TokenStream {
-    if is_async {
-        quote! { #[async_std::test] }
+fn resolve_test_attr(
+    is_async: bool,
+    explicit_test_attr: Option<TokenStream>,
+    attributes: &[Attribute],
+    error_span: Span,
+) -> syn::Result<Option<TokenStream>> {
+    if !is_async {
+        Ok(Some(quote! { #[test] }))
+    } else if let Some(explicit_attr) = explicit_test_attr {
+        Ok(Some(explicit_attr))
+    } else if attributes
+        .iter()
+        .any(|attr| attr_ends_with(attr, &parse_quote! {test}))
+    {
+        // test attr is already in the attributes; we don't need to re-inject it
+        Ok(None)
     } else {
-        quote! { #[test] }
+        Err(syn::Error::new(error_span, "async tests require either an explicit `test_attr` or an attribute whose path ends with `test`"))
```

Comment:
You can use a `compiler_error!` macro instead : you're in the rendering stage so it's legit to use it to inform the user.

Anyway, maybe the best thing is to inspect the data after parsing and catch the case where the mandatory test attribute is missed.

Reply:
Yeah, I know the difference, this is way I prefer to catch all errors in parsing stage or by inspecting the parsed output to check coherence.  Take a look to `error::rstest` to see how all other cases are handled.